mlflow/tracing/types/wrapper.py

```diff
@@ -106,14 +103,32 @@ def set_attributes(self, attributes: Dict[str, Any]):
                 f"Attributes must be a dictionary, but got {type(attributes)}. Skipping."
             )
             return
-        self._attributes.update(attributes)
+
+        for key, value in attributes.items():
+            self.set_attribute(key, value)
 
     def set_attribute(self, key: str, value: Any):
         """Set a single attribute to the span."""
         if not isinstance(key, str):
             _logger.warning(f"Attribute key must be a string, but got {type(key)}. Skipping.")
             return
-        self._attributes[key] = value
+
+        # NB: OpenTelemetry attribute can store not only string but also a few primitives like
+        #   int, float, bool, and list of them. However, we serialize all into JSON string here
+        #   for the simplicity in deserialization process.
+        self._span.set_attribute(key, json.dumps(value, cls=TraceJSONEncoder))
+
```

Comment:
@B-Step62 Looks like this results in lots of escape sequences (`\`) throughout the attributes, which then renders as double quotes in the UI.

<img width="582" alt="Screenshot 2024-04-15 at 12 22 54 PM" src="https://github.com/mlflow/mlflow/assets/39497902/988bce17-ff62-4ee1-9840-647465ee878f">

Can we only dump things that aren't already OTel-supported value types to JSON?

Reply:
@dbczumar 

The UI shows unnecessary escape because it doesn't do JSON deserialization now. It will be removed once updated.

> Can we only dump things that aren't already OTel-supported value types to JSON?

It's tricky because once we dumps sth into JSON we can't know if it was serialized by us or was a string from the beginning.

Reply:
Does this work?:

When a user sets an attribute, check if it's of type str, int, float, or bool. If it is, don't stringify it. If it isn't one of these types, stringify it.

Reply:
When frontend recieves str (or when we load json back to Python object), how do we know whether WE did the stringification or user?

Reply:
Discussed offline: We decided to do JSON serialization for all attribute values to be compatible with [OpenTelemetry spec](https://opentelemetry.io/docs/concepts/signals/traces/#attributes) when saving spans in blob storage.